website/registrations/utils.py

```diff
@@ -0,0 +1,376 @@
+        self.log_error = functools.partial(log_error, type=self.get_field_type())
+
+    def get_field_type(self):
+        return self.format if self.type == 'choose' else self.type
+
+    def _validate(self):
+        parsed_value = None
+        if self.required and not bool(self.value):
+            self.log_error(missing=True)
+        else:
+            if self.type == 'string':
+                parsed_value = self.value
+            elif self.type == 'choose' and self.format in ['singleselect', 'multiselect']:
+                if self.format == 'singleselect':
+                    if self.value not in self.options:
+                        self.log_error(invalid=True)
+                    else:
+                        parsed_value = self.value
+                else:
+                    parsed_value = []
+                    choices = [val.strip() for val in self.value.split(';')]
+                    for choice in choices:
+                        if choice not in self.options:
+                            self.log_error(invalid=True)
+                        else:
+                            parsed_value.append(choice)
+            self._parsed_value = parsed_value
+
+class MetadataField(UploadField):
+    def __init__(self, value, validations, log_error):
+        super(MetadataField, self).__init__()
+        self.format = validations.get('format', 'string')
+        self.required = validations.get('required', False)
+        self.value = value.strip()
+        self.log_error = log_error
+
+    def get_field_type(self):
+        return self.format
+
+    def _validate(self):
+        parsed_value = None
+        if self.required and not bool(self.value):
+            self.log_error(missing=True)
+        else:
+            if self.format == 'string':
+                parsed_value = self.value
+            elif self.format == 'list':
+                parsed_value = [val.strip() for val in self.value.split(';')]
+            self._parsed_value = parsed_value
+
+class ContributorField(MetadataField):
+    # format: contributor_name<contributor_email>;contributor_name<contributor_email>
+    contributor_regex = re.compile(r'(?P<full_name>[\w -]+)<(?P<email>.*?)>')
+    def _validate(self):
+        parsed_value = None
+        if self.required and not bool(self.value):
+            self.log_error(missing=True)
+        else:
+            parsed_value = []
+            parsed_contributor_list = [val.strip() for val in self.value.split(';')]
+            for contrib in parsed_contributor_list:
+                match = self.contributor_regex.match(contrib.strip())
+                if match:
+                    try:
+                        full_name = match.group('full_name')
+                        email = match.group('email')
+                    except AttributeError:
+                        self.log_error(invalid=True)
+                    else:
+                        parsed_value.append({'full_name': full_name, 'email': email})
```

Comment:
Minor nitpicking: do we need to detect dup contributors?
* If done in the parser, then we can use `set` as the type for `parsed_value`. However this requires sending them a success but warning modal. However, set is much slower during iteration.
* If done in the creation task, dups can be ignored. I can include the warning in the success emails.

Reply:
Probably don't need to detect duplicates, since they will be ignored. I don't know whether use cares about dups either as long as the dups gets added as a contributor once.